pkg/compactv2/modifiers.go

```diff
@@ -0,0 +1,334 @@
+	"github.com/prometheus/prometheus/tsdb/index"
+	"github.com/prometheus/prometheus/tsdb/tombstones"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+)
+
+type Modifier interface {
+	Modify(sym index.StringIter, set storage.ChunkSeriesSet, log ChangeLogger, p ProgressLogger) (index.StringIter, storage.ChunkSeriesSet)
+}
+
+type DeletionModifier struct {
+	deletions []metadata.DeletionRequest
+}
+
+var (
+	SupportedRelabelActions = map[relabel.Action]struct{}{relabel.Keep: {}, relabel.Drop: {}}
+)
+
+func WithDeletionModifier(deletions ...metadata.DeletionRequest) *DeletionModifier {
+	return &DeletionModifier{deletions: deletions}
+}
+
+func (d *DeletionModifier) Modify(sym index.StringIter, set storage.ChunkSeriesSet, log ChangeLogger, p ProgressLogger) (index.StringIter, storage.ChunkSeriesSet) {
+	// TODO(bwplotka): Modify symbols as well. Otherwise large string will be kept forever.
+	// This is however what Prometheus already does. It does not increase index size too much though.
+	// This needs a bit of work due to sorting and tracking required to rebuild them.pp
+
+	return sym, &delModifierSeriesSet{
+		d: d,
+
+		ChunkSeriesSet: set,
+		log:            log,
+		p:              p,
+	}
+}
+
+type delModifierSeriesSet struct {
+	storage.ChunkSeriesSet
+
+	d   *DeletionModifier
+	log ChangeLogger
+	p   ProgressLogger
+
+	curr *storage.ChunkSeriesEntry
+	err  error
+}
+
+func (d *delModifierSeriesSet) Next() bool {
+SeriesLoop:
+	for d.ChunkSeriesSet.Next() {
+		s := d.ChunkSeriesSet.At()
+		lbls := s.Labels()
+
+		var intervals tombstones.Intervals
+		for _, deletions := range d.d.deletions {
+			for _, m := range deletions.Matchers {
+				v := lbls.Get(m.Name)
+				if v == "" {
+					continue
+				}
+
+				if !m.Matches(v) {
+					continue
+				}
+				if len(deletions.Intervals) > 0 {
+					for _, in := range deletions.Intervals {
+						intervals = intervals.Add(in)
+					}
+					break
+				}
+
+				// Special case: Delete whole series.
+				chksIter := s.Iterator()
+				var chks []chunks.Meta
+				for chksIter.Next() {
+					chks = append(chks, chksIter.At())
+				}
+				if d.err = chksIter.Err(); d.err != nil {
+					return false
+				}
+
+				var deleted tombstones.Intervals
+				if len(chks) > 0 {
+					deleted = deleted.Add(tombstones.Interval{Mint: chks[0].MinTime, Maxt: chks[len(chks)-1].MaxTime})
+				}
+				d.log.DeleteSeries(lbls, deleted)
+				d.p.SeriesProcessed()
+				continue SeriesLoop
+			}
+		}
+
+		d.curr = &storage.ChunkSeriesEntry{
+			Lset: lbls,
+			ChunkIteratorFn: func() chunks.Iterator {
+				return NewDelGenericSeriesIterator(s.Iterator(), intervals, func(intervals tombstones.Intervals) {
+					d.log.DeleteSeries(lbls, intervals)
+				}).ToChunkSeriesIterator()
+			},
+		}
+		return true
+	}
+	return false
+}
+
+// Intersection returns intersection between interval and range of intervals.
+func Intersection(i tombstones.Interval, dranges tombstones.Intervals) tombstones.Intervals {
```

Comment:
Seems this is only used in this package, can we make it private?

Reply:
yes, good point. Thanks!